specs/_features/whisk/beacon-chain.md

```diff
@@ -0,0 +1,490 @@
+        assert is_k_commitment_unique(state, body.whisk_k_commitment)
+        assert IsValidWhiskOpeningProof(
+            body.whisk_tracker,
+            body.whisk_k_commitment,
+            body.whisk_registration_proof,
+        )
+        proposer.whisk_tracker = body.whisk_tracker
+        proposer.whisk_k_commitment = body.whisk_k_commitment
+    else:  # next Whisk proposals
+        assert body.whisk_registration_proof == WhiskTrackerProof()
+        assert body.whisk_tracker == WhiskTracker()
+        assert body.whisk_k_commitment == BLSG1Point()
+    assert body.whisk_shuffle_proof_M_commitment == BLSG1Point()
+```
+
+```python
+def process_block(state: BeaconState, block: BeaconBlock) -> None:
+    process_block_header(state, block)
+    if is_execution_enabled(state, block.body):
+        process_withdrawals(state, block.body.execution_payload)
+        process_execution_payload(state, block.body.execution_payload, EXECUTION_ENGINE)
+    process_randao(state, block.body)
+    process_eth1_data(state, block.body)
+    process_operations(state, block.body)
+    process_sync_aggregate(state, block.body.sync_aggregate)
+    process_whisk(state, block.body)  # [New in Whisk]
+```
+
+### Deposits
+
+```python
+def get_unique_whisk_k(state: BeaconState, validator_index: ValidatorIndex) -> BLSFieldElement:
+    counter = 0
+    while True:
+        # hash `validator_index || counter`
+        k = BLSFieldElement(bytes_to_bls_field(hash(uint_to_bytes(validator_index) + uint_to_bytes(uint64(counter)))))
+        if is_k_commitment_unique(state, BLSG1ScalarMultiply(k, BLS_G1_GENERATOR)):
+            return k  # unique by trial and error
+        counter += 1
+```
+
+```python
+def get_initial_commitments(k: BLSFieldElement) -> Tuple[BLSG1Point, WhiskTracker]:
+    return (
+        BLSG1ScalarMultiply(k, BLS_G1_GENERATOR),
+        WhiskTracker(r_G=BLS_G1_GENERATOR, k_r_G=BLSG1ScalarMultiply(k, BLS_G1_GENERATOR))
+    )
+```
+
+```python
+def get_validator_from_deposit_whisk(
+    state: BeaconState,
+    pubkey: BLSPubkey,
+    withdrawal_credentials: Bytes32,
+    amount: uint64
+) -> Validator:
+    effective_balance = min(amount - amount % EFFECTIVE_BALANCE_INCREMENT, MAX_EFFECTIVE_BALANCE)
+    k = get_unique_whisk_k(state, ValidatorIndex(len(state.validators)))
+    whisk_k_commitment, whisk_tracker = get_initial_commitments(k)
+
+    validator = Validator(
+        pubkey=pubkey,
+        withdrawal_credentials=withdrawal_credentials,
+        activation_eligibility_epoch=FAR_FUTURE_EPOCH,
+        activation_epoch=FAR_FUTURE_EPOCH,
+        exit_epoch=FAR_FUTURE_EPOCH,
+        withdrawable_epoch=FAR_FUTURE_EPOCH,
+        effective_balance=effective_balance,
+        # Whisk fields
+        whisk_tracker=whisk_tracker,
+        whisk_k_commitment=whisk_k_commitment,
+    )
+    return validator
+```
+
+```python
+def process_deposit(state: BeaconState, deposit: Deposit) -> None:
```

Comment:
Hmm, I dont quite understand the entirety of #3177. What does this mean for Whisk and how can we fix this?

Reply:
Oh sorry @hwwhww missed the update, pushed a commit to extend altair's apply_deposit https://github.com/ethereum/consensus-specs/pull/3342/commits/1ea91d403f70b7c4d8d731b33fdc7bb540880b74

- Would be cool to have this change in to prevent such oversights https://github.com/ethereum/consensus-specs/pull/3311